### DISCUSSION TITLE
How to draw a stress-strain curve from a uni-axial traction simulation

### DISCUSSION THREAD

**Discussion by Xelver** [idaholab/moose#18241] [Q&A Modules: Solid mechanics] [answered]
Hi, guys,

I'm working on a simple uniaxial traction simulation for a 1mm x 1mm x 1mm cube with some randomly distributed defect elements inside, as:

![image](https://user-images.githubusercontent.com/74576082/124248379-5e8e0180-db55-11eb-9873-ee0379805d4e.png)

Because there are some defects, the stress and strain in different elements are not the same. I've tried to get the average value on the surface or the whole cube, but the result seems wrong.
In Abaqus, we can set a reference point and couple the point with the traction surface. Then we can obtain the displacement and the reaction force of the reference point to calculate the stress-strain curve.  So I am wondering whether MOOSE has a similar function.

Thanks for your answers

**Comment by WilkAndy** [comment]
What do you mean "the result seems wrong"?  Usually you plot average stress vs average strain (which you can get using a couple of Postprocessors).  The average stress is equal to (average) surface traction, so you could use that instead.  The average strain is equal to (average) surface displacement, so you could use that instead.   Obviously you have to choose the relevant components of stress and strain.

Perhaps the result "looks wrong" because you're incorrectly normalising using the void volume?

**Reply by Xelver** [reply]
Thanks for your answer. 
I had used **ElementAverageValue** to get average stress and average strain, but the curve I had got was exactly the same as the cube without the defects. (Normally the slope should be lower).  I think it's because the **ElementAverageValue** doesn't count the volume of the defect. 
So I used another method.  I didn't delete the defect elements, instead, I put them into a subdomain and assigned them a very small Young's modules（0.001） and poisons ratio （1e-11）.  The slope changed with this method, but it still wasn't what I wanted.  Normally the Young's modulus of a cube (430GPa, 0.2) with 10% voids is about 352 GPa ( I've got the same result with ABAQUS), but here I've got 370 GPa.

**Reply by WilkAndy** [reply]
Firstly, i'd check your boundary conditions - load the result into paraview and check that the cube "moves" in roughly the same way as ABAQUS.  You're getting a stiffer result, so maybe the MOOSE model is more constrained than the ABAQUS one.

Secondly, you could easily be using different shape functions in ABAQUS and MOOSE.  If you use linear-lagrange (the default in MOOSE) then you'll get a stiffer result because the solid material can't deform accurately around a 1-element void.  That's a bit hard to explain - tell me if you don't understand.

You're correct about the `ElementAverageValue` : it simply averages over the elements.

**Reply by Xelver** [reply]
The boundary conditions are 
- disp_x = 0, left
- disp_y = 0, bottom
- disp_z = 0, back
- disp_y = f, top (1e-6 m at 10 s)
I've used the same boundary conditions for ABAQUS.

Yes, I used the default option as follows.  I'll try to use higher-order later. But I'm not very clear about it. Could you please explain a little bit if it doesn't trouble you?
```
[Modules/TensorMechanics/Master]
  [./all]
    add_variables = true
    strain = FINITE
    #incremental = true
    generate_output = 'vonmises_stress stress_yy strain_yy'
    use_displaced_mesh = true
  [../]
[]
```

**Reply by Xelver** [reply]
Hey, I have tried using a higher-order `elem_type` with HEX20 and HEX27 with second-order-Lagrange variables. The resulted effective Young's modulus are 355 GPa (which is closer to ABAQUS result)  for HEX20 and 348 GPa for HEX27.

**Comment by ngrilli** [comment, marked as answer]
Dear @Xelver 

It's easy to plot stress-strain curves with paraview:
1) Click on "Select cell through" (or "f" on keyboard)
2) Select the element at the boundary where you want the stress
3) Filters > Extract Selection > Apply
Now you will see only the selected elements
4) Filters > Integrate Variables + tick "Divide cell data by volume" in the properties tab > Apply
In the spreadsheet on the right you will see the average quantities, select Attribute > Cell Data
and you will see your average stress.
If you want to plot the stress as a function of time, you can select the line in the spreadsheet,
then Filters > Plot Selection over time
You will see all Cell Quantities as a function of time, averaged over the selected element.
Because you can see the selected elements, making mistakes is very unlikely.

Trust this helps,
Nicolò

**Reply by Xelver** [reply]
Thanks for your answer. 
It really helps. But I want to obtain the curve of the entire object or the "equivalent young's modulus" of the object . 
I used your method to get the average value of the cube, the results were as same as using postprocessors **ElementAverageValue**.  So I thought that it may not be the average value that we should use to draw the stress-strain curve?

**Reply by ngrilli** [reply]
Dear @Xelver 

Getting a stress-strain curve by averaging over all elements does not make a lot of sense to me.
Physically the stress is measured as average over a surface.
I am not sure what you mean by "equivalent Young's modulus",
if your material properties are homogeneous, the Young modulus will be the same everywhere anyway.

If you mean you want an effective Young modulus of the structure, which will be lower than Young modulus
of the bulk because of the holes, then definitely average the stress on a surface.
You can repeat the operations on different surfaces and with different distribution of the holes
to get more statistics

Nicolò

**Reply by Xelver** [reply]
Dear @ngrilli 

I think that you're right.  I mean the effective Young modulus of the structure. 
So I tried to get the stress on different surfaces with different distributions of the holes.  The boundary conditions are

disp_x = 0, left
disp_y = 0, bottom
disp_z = 0, back
disp_y = f, top (1e-6 m at 10 s)

The average stress of the bottom and the top kept pretty much unchanged with different distribution of the holes, while the average stress of the other surfaces changed a lot with the different distribution.  So I'm not sure which could be used to calculate the effective Young's modulus.

**Comment by sapitts** [comment]
Hi @Xelver,

If you would like to determine the effective Young's modulus of the complete structure, you may also find the Asymptotic Expansion Homogenization capabilities useful, particularly the [AsymptoticExpansionHomogenizationElasticConstants postprocessor](https://mooseframework.inl.gov/source/postprocessors/AsymptoticExpansionHomogenizationElasticConstants.html) and the associated kernels--links to documentation on the kernels are given on the postprocessor page. More details about this approach are available in [Hales et al.(2015)](https://doi.org/10.1016/j.commatsci.2014.12.039).
